{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "71c909f7-1349-49c7-ba27-08f81922b489",
   "metadata": {},
   "outputs": [],
   "source": [
    "%config InlineBackend.figure_format = 'svg'\n",
    "import re\n",
    "import json\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import cooler\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.colors as colors\n",
    "import chromosight.utils.contacts_map as mapp\n",
    "from IPython.display import display\n",
    "import chromosight.utils.io as io\n",
    "from scipy.signal import find_peaks\n",
    "from scipy import ndimage as ndi\n",
    "import matplotlib.pyplot as plt\n",
    "from skimage.feature import peak_local_max\n",
    "from skimage.transform import resize\n",
    "from skimage import data, img_as_float"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "1e8a0994-9af8-4753-aa4a-02a8197ea678",
   "metadata": {},
   "outputs": [],
   "source": [
    "#path: The .cool file path input(if use .mcool file rather than .cool file, need to choose \n",
    "#resolution by add \"::/resolutions/{resolution value} \" after path, resolutions value can only\n",
    "#choose from {1000,5000,10000,25000,50000,100000,250000,500000,1000000,2500000}, The smaller \n",
    "#the value, the more detailed the matrix and the image.ï¼‰\n",
    "class heatmap_anylsis:\n",
    "    \n",
    "    def __init__(self,path):\n",
    "        self.path = path\n",
    "        \n",
    "#use cooler package generate cooler api  \n",
    "# will return cooler.api.cooler \n",
    "    def readcool(self):\n",
    "        new_cool = cooler.Cooler(self.path)\n",
    "        return new_cool\n",
    "    \n",
    "## Converting imported .cool files into two-dimensional arrays    \n",
    "    def getmatrix(self,value = False):\n",
    "        new_cool = cooler.Cooler(self.path)\n",
    "        new_matrix = new_cool.matrix(sparse=False, balance=value)[:]\n",
    "        return new_matrix\n",
    "    \n",
    "    \n",
    "    def generatemap(self,matrix,size = (8,8)):\n",
    "        plt.figure(figsize=size)\n",
    "        plt.imshow(matrix ** 0.02, cmap='afmhot_r',norm=colors.PowerNorm(gamma=0.45))\n",
    "\n",
    "    \n",
    "## generate heatmap from matrix(two-dimensional arrays)  \n",
    "    def generatemap2(self,matrix,size = (8,8)):\n",
    "        plt.figure(figsize=size)\n",
    "        if matrix.shape[0] < 200:\n",
    "            matrix_c = matrix.copy()\n",
    "            matrix_c = resize(matrix_c, (250,matrix.shape[1]))\n",
    "            plt.imshow(matrix_c ** 0.02, cmap='afmhot_r',norm=colors.PowerNorm(gamma=0.45))\n",
    "            plt.axis('off')\n",
    "        #resize unvisible graph\n",
    "        else:\n",
    "            plt.imshow(matrix ** 0.02, cmap='afmhot_r',norm=colors.PowerNorm(gamma=0.45))\n",
    "            plt.axis('off')\n",
    "    \n",
    "## Table containing bin related informations(start,end position and nomalization information)        \n",
    "    def bins_info(self):\n",
    "        all_info = mapp.HicGenome(self.path)\n",
    "        bin_info = all_info.bins\n",
    "        return bin_info\n",
    "    \n",
    "## Table of chromosome information. Each row contains the name, length, first and last bin of a chromosome.    \n",
    "    def chro_info(self):\n",
    "        table_c = io.load_cool(self.path)\n",
    "        return table_c[1]\n",
    "    \n",
    "## thransfer bin start and end location from dataframe to list  \n",
    "    def thansfer_bin_to_list(self,list_type='start_list'):\n",
    "        choro_frame = self.chro_info()\n",
    "        bin_loc = choro_frame.iloc[:,2:4]\n",
    "        #extract start and end bin column\n",
    "        start_bin = bin_loc.iloc[:,0]\n",
    "        end_bin = bin_loc.iloc[:,1]\n",
    "        #transfer to list\n",
    "        start_list = start_bin.values.tolist()\n",
    "        end_list = end_bin.values.tolist()\n",
    "        if list_type == 'start_list':\n",
    "            return start_list\n",
    "        elif list_type == 'end_list':\n",
    "            return end_list\n",
    "        else:\n",
    "            print('Please input \"start_list\" or \"end_list\" to choose list type')\n",
    "            \n",
    "## split matrix by each contig row(use range of bins i.e. range from start bin to end bin, generate each contig row matrix, also will return a dictionary\n",
    "## related contig id to matrix\n",
    "    def split_matrix(self,rtype = 'list'):\n",
    "        n_list=[]\n",
    "        n_dic = {}\n",
    "        chro_info = self.chro_info()\n",
    "        #extract contig id column and thansfer to list\n",
    "        z = chro_info.iloc[:,0]\n",
    "        z = z.values.tolist()\n",
    "        start_list = self.thansfer_bin_to_list()\n",
    "        end_list =self.thansfer_bin_to_list('end_list')\n",
    "        image = self.getmatrix()\n",
    "        for i in range(len(start_list)):\n",
    "            #get each contig start bin and end bin position,extract each contig row\n",
    "            row = image[start_list[i]:end_list[i],:]\n",
    "            n_dic[z[i]] = row\n",
    "            n_list.append(row)\n",
    "        if rtype == 'list':\n",
    "            return n_list\n",
    "        elif rtype == 'dic':\n",
    "            return n_dic\n",
    "        \n",
    "## remove array from list       \n",
    "    def removearray(self,List,arr):\n",
    "        ind = 0\n",
    "        size = len(List)\n",
    "        while ind != size and not np.array_equal(List[ind],arr):\n",
    "            ind += 1\n",
    "        if ind != size:\n",
    "            List.pop(ind)\n",
    "        else:\n",
    "            raise ValueError('array not found in list.')\n",
    "            \n",
    "    def get_contig_id(self,matrix,my_dict):\n",
    "        for key, value in my_dict.items():\n",
    "            if np.array_equal(value,matrix):\n",
    "                return key\n",
    " \n",
    "        return \"key doesn't exist\"\n",
    "\n",
    "## get contig id from matrix that interested\n",
    "    def sort_matrix(self):\n",
    "        matrix_list = self.split_matrix()\n",
    "        m_list = []\n",
    "        c_list = matrix_list.copy()\n",
    "        #put width info into m_list,sort matrix by matrix width\n",
    "        for i in matrix_list:\n",
    "            m_list.append(i.shape[0])\n",
    "            m_list.sort()\n",
    "        sorted_list = []\n",
    "        #accoring m_list find 2D matrix\n",
    "        for i in m_list:\n",
    "            for j in c_list:\n",
    "                if j.shape[0] == i:\n",
    "                    sorted_list.append(j)\n",
    "                    self.removearray(c_list,j)\n",
    "                    break    \n",
    "        return sorted_list\n",
    "    \n",
    "    def remove_contig_length(self,length):\n",
    "        chrom_info = self.chro_info() \n",
    "        splited_matrix = self.split_matrix()\n",
    "        #extract contig length column and thansfer to list\n",
    "        length_list = chrom_info.iloc[:,1].values.tolist() #length list\n",
    "        list1 = []\n",
    "        for i in length_list:\n",
    "            if i >=length: \n",
    "                list1.append(i)#list 1 contain all length in the list wich longer than length value\n",
    "        list_index = []\n",
    "        for i in range(len(length_list)):\n",
    "            for j in list1:\n",
    "                if length_list[i] == j:\n",
    "                    list_index.append(i)# get contig length index which longer than length value\n",
    "        list_new = []\n",
    "        for i in list_index:\n",
    "            list_new.append(splited_matrix[i])#according index find contig matrix\n",
    "        return list_new\n",
    "    \n",
    "## show peak diagram for contig matrix(cannot use this function to whole heatmap)\n",
    "    def peak_dia(self,contig_matrix):\n",
    "        plt.plot(contig_matrix.T)\n",
    "        plt.show()\n",
    "        \n",
    "    def find_location(self,p_index,dic,matrix,choro_frame):\n",
    "        bin_loc = choro_frame.iloc[:,2:4] #get start_bin and end_bin dataframe\n",
    "        start_bin = bin_loc.iloc[:,0]# change to list\n",
    "        end_bin = bin_loc.iloc[:,1]# change to list\n",
    "        for i in range(len(start_bin)):\n",
    "            if p_index >= start_bin[i] and p_index < end_bin[i]:  #  if input index larger or equel to start bin and less than end bin         \n",
    "                idd = self.get_contig_id(matrix[start_bin[i]:end_bin[i],:],dic)#extract 2D array betwwen start_bin and end_bin and use dictionary find contig id\n",
    "        return idd\n",
    "    \n",
    "    "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "hraaxl_chromo",
   "language": "python",
   "name": "hraaxl_chromo"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
